// satteri/dist/hast/hast-visitor.js
import { materializeHastNode } from "./hast-materializer.js";
import { HastReader, HAST_ELEMENT, HAST_TEXT, HAST_COMMENT, HAST_RAW, HAST_MDX_JSX_ELEMENT, HAST_MDX_JSX_TEXT_ELEMENT, HAST_MDX_FLOW_EXPRESSION, HAST_MDX_TEXT_EXPRESSION, HAST_MDX_ESM, } from "./hast-reader.js";
import { CommandBuffer } from "../command-buffer.js";
import { walkHandle, applyCommandsToHandle, serializeHandle, textContentHandle, getNodeData as napiGetNodeData, parseExpression as napiParseExpression, parseEsm as napiParseEsm, } from "#binding";
/** Maps HastNode objects to their arena node IDs without Object.defineProperty overhead. */
const nodeIdMap = new WeakMap();
/** Attach `parseExpression()` to an MDX expression or ESM node. */
function attachParseExpression(node, parseFn) {
    Object.defineProperty(node, "parseExpression", {
        value() {
            const value = this.value;
            if (typeof value !== "string")
                return null;
            const json = parseFn(value);
            if (json == null)
                return null;
            return JSON.parse(json);
        },
        writable: false,
        enumerable: false,
        configurable: true,
    });
}
/**
 * Serialize a HastNode for the command buffer. Marks each node `_hast: true`,
 * and — except at the root, which is the new replacement content — emits a
 * reused (materialized) node as a `{ _ref: id }` placeholder so the rebuild
 * splices the original in place (preserving its id, applying any pending patch
 * on it) instead of rebuilding it fresh.
 */
function markHast(node) {
    return markHastNode(node, true);
}
function markHastNode(node, isRoot) {
    if (!isRoot) {
        const id = nodeIdMap.get(node) ?? node._nodeId;
        if (typeof id === "number")
            return { _ref: id };
    }
    const obj = { _hast: true, type: node.type };
    if ("tagName" in node)
        obj.tagName = node.tagName;
    if ("properties" in node)
        obj.properties = node.properties;
    if ("value" in node)
        obj.value = node.value;
    if ("name" in node)
        obj.name = node.name;
    if ("attributes" in node)
        obj.attributes = node.attributes;
    if ("data" in node && node.data != null)
        obj.data = node.data;
    if ("children" in node) {
        obj.children = node.children.map((c) => markHastNode(c, false));
    }
    return obj;
}
function nid(node) {
    return nodeIdMap.get(node) ?? node._nodeId;
}
function asArray(value) {
    return Array.isArray(value) ? value : [value];
}
class HastVisitorContextImpl {
    #commandBuffer = new CommandBuffer();
    #diagnostics = [];
    /** Track accumulated node state for multiple setProperty calls on the same node. */
    #pendingNodes = new Map();
    #handle;
    #getSource;
    fileURL;
    constructor(handle, getSource, fileURL) {
        this.#handle = handle;
        this.#getSource = getSource;
        this.fileURL = fileURL;
    }
    get source() {
        const value = this.#getSource();
        Object.defineProperty(this, "source", { value, writable: false, enumerable: true });
        return value;
    }
    removeNode(node) {
        this.#commandBuffer.removeNode(nid(node));
    }
    replaceNode(node, newNode) {
        const id = nid(node);
        this.#commandBuffer.replaceRawJson(id, JSON.stringify(markHast(newNode)));
        this.#pendingNodes.set(id, newNode);
    }
    insertBefore(node, newNode) {
        const id = nid(node);
        for (const n of asArray(newNode)) {
            this.#commandBuffer.insertBeforeRawJson(id, JSON.stringify(markHast(n)));
        }
    }
    insertAfter(node, newNode) {
        const id = nid(node);
        for (const n of asArray(newNode)) {
            this.#commandBuffer.insertAfterRawJson(id, JSON.stringify(markHast(n)));
        }
    }
    wrapNode(node, parentNode) {
        this.#commandBuffer.wrapNodeRawJson(nid(node), JSON.stringify(markHast(parentNode)));
    }
    prependChild(node, childNode) {
        const id = nid(node);
        for (const n of asArray(childNode)) {
            this.#commandBuffer.prependChildRawJson(id, JSON.stringify(markHast(n)));
        }
    }
    appendChild(node, childNode) {
        const id = nid(node);
        for (const n of asArray(childNode)) {
            this.#commandBuffer.appendChildRawJson(id, JSON.stringify(markHast(n)));
        }
    }
    insertChildAt(node, index, childNode) {
        const children = "children" in node ? node.children : [];
        if (index <= 0 || children.length === 0) {
            this.prependChild(node, childNode);
        }
        else if (index >= children.length) {
            this.appendChild(node, childNode);
        }
        else {
            this.insertBefore(children[index], childNode);
        }
    }
    removeChildAt(node, index) {
        const child = "children" in node ? node.children[index] : undefined;
        if (child)
            this.removeNode(child);
    }
    setProperty(node, key, value) {
        const id = nid(node);
        if (key === "children") {
            // children is structural: set-children keeps the node and swaps only its
            // child list (reused children keep their id).
            const wrapper = {
                _hast: true,
                type: "root",
                children: value.map((child) => markHastNode(child, false)),
            };
            this.#commandBuffer.setChildren(id, JSON.stringify(wrapper));
            return;
        }
        if (key === "data") {
            this.#commandBuffer.setProperty(id, key, value != null ? JSON.stringify(value) : null);
            return;
        }
        if (node.type === "element") {
            // Fast binary path, no materialization, no JSON serialization
            this.#commandBuffer.setProperty(id, key, value);
            return;
        }
        if (node.type === "mdxJsxFlowElement" || node.type === "mdxJsxTextElement") {
            // MDX JSX nodes use `attributes`, not `properties`, keep replaceNode path
            const pending = this.#pendingNodes.get(id);
            const current = (pending ?? node);
            const updated = { ...current };
            const attrs = [...(updated.attributes ?? [])];
            const idx = attrs.findIndex((a) => a.type === "mdxJsxAttribute" && a.name === key);
            if (idx !== -1)
                attrs.splice(idx, 1);
            const attrValue = value === true || value === null || value === undefined
                ? null
                : typeof value === "string"
                    ? value
                    : String(value);
            attrs.push({ type: "mdxJsxAttribute", name: key, value: attrValue });
            updated.attributes = attrs;
            this.replaceNode(node, updated);
            return;
        }
        // Text-like nodes (text, comment, raw, expressions, esm), fast binary path.
        // Rust handles "value" setProperty directly on these types.
        this.#commandBuffer.setProperty(id, key, value);
    }
    textContent(node) {
        return textContentHandle(this.#handle, nid(node));
    }
    report({ message, node, severity = "error", }) {
        this.#diagnostics.push({ message, nodeId: node ? nid(node) : undefined, severity });
    }
    getCommandBuffer() {
        return this.#commandBuffer;
    }
    getDiagnostics() {
        return this.#diagnostics;
    }
}
function isFilteredVisitor(v) {
    return typeof v === "object" && v !== null && "filter" in v && "visit" in v;
}
/** Node types that use filtered visitors (have tag/component names). */
const FILTERED_METHODS = new Set(["element", "mdxJsxFlowElement", "mdxJsxTextElement"]);
/** Resolve subscriptions from a plugin instance. */
export function resolveSubscriptions(plugin) {
    const subs = [];
    for (const [methodName, nodeType] of Object.entries(METHOD_TO_TYPE)) {
        const value = plugin[methodName];
        if (value === undefined)
            continue;
        if (FILTERED_METHODS.has(methodName)) {
            const items = Array.isArray(value) ? value : [value];
            for (const fv of items) {
                subs.push({
                    nodeType,
                    tagFilter: fv.filter,
                    visitFn: fv.visit,
                });
            }
        }
        else {
            // Bare function, empty filter matches all nodes of this type
            subs.push({ nodeType, tagFilter: [], visitFn: value });
        }
    }
    return subs;
}
/** Reverse map: method name → node type number */
const METHOD_TO_TYPE = {
    element: HAST_ELEMENT,
    text: HAST_TEXT,
    comment: HAST_COMMENT,
    raw: HAST_RAW,
    doctype: 4, // HAST_DOCTYPE
    mdxJsxFlowElement: HAST_MDX_JSX_ELEMENT,
    mdxJsxTextElement: HAST_MDX_JSX_TEXT_ELEMENT,
    mdxFlowExpression: HAST_MDX_FLOW_EXPRESSION,
    mdxTextExpression: HAST_MDX_TEXT_EXPRESSION,
    mdxjsEsm: HAST_MDX_ESM,
};
/**
 * Selective walk path: Rust walks the tree, only sends matched nodes to JS.
 */
const textDecoder = new TextDecoder("utf-8");
/** Decode properties from the walk buffer at the given position. */
function decodeProperties(view, buf, pos) {
    const propCount = view.getUint16(pos, true);
    pos += 2;
    const properties = {};
    for (let i = 0; i < propCount; i++) {
        const nameLen = view.getUint16(pos, true);
        pos += 2;
        const name = textDecoder.decode(buf.subarray(pos, pos + nameLen));
        pos += nameLen;
        const kind = buf[pos];
        pos += 1;
        const valLen = view.getUint16(pos, true);
        pos += 2;
        const valStr = textDecoder.decode(buf.subarray(pos, pos + valLen));
        pos += valLen;
        switch (kind) {
            case 0: // PROP_STRING
                properties[name] = valStr;
                break;
            case 1: // PROP_BOOL_TRUE
                properties[name] = true;
                break;
            case 2: // PROP_BOOL_FALSE
                break;
            case 3: // PROP_SPACE_SEP
                properties[name] = valStr.split(" ").filter((s) => s.length > 0);
                break;
            case 5: // PROP_INT
                properties[name] = Number(valStr);
                break;
        }
    }
    return properties;
}
/** Decode the 24-byte position block written by `serialize_hast_node_inline`. */
function readPositionPrefix(view, offset) {
    const startOffset = view.getUint32(offset, true);
    const endOffset = view.getUint32(offset + 4, true);
    const startLine = view.getUint32(offset + 8, true);
    const startColumn = view.getUint32(offset + 12, true);
    const endLine = view.getUint32(offset + 16, true);
    const endColumn = view.getUint32(offset + 20, true);
    if (startLine === 0 && startOffset === 0)
        return undefined;
    return {
        start: { offset: startOffset, line: startLine, column: startColumn },
        end: { offset: endOffset, line: endLine, column: endColumn },
    };
}
/**
 * Walk-path element node: uses prototype getters instead of per-instance
 * Object.defineProperty. V8 optimises shared hidden classes far better,
 * this is ~16x faster for construction than the defineProperty approach.
 *
 * The buffer reference data is stored on private instance fields so the
 * prototype getter can decode lazily on first access.
 */
class WalkElement {
    type = "element";
    /** @internal */ _view;
    /** @internal */ _buf;
    /** @internal */ _propsPos;
    /** @internal */ _childIds;
    /** @internal */ _resolver;
    get properties() {
        const val = decodeProperties(this._view, this._buf, this._propsPos);
        Object.defineProperty(this, "properties", {
            value: val,
            writable: true,
            enumerable: true,
            configurable: true,
        });
        return val;
    }
    get children() {
        const val = this._resolver.materializeChildren(this._childIds);
        Object.defineProperty(this, "children", {
            value: val,
            writable: true,
            enumerable: true,
            configurable: true,
        });
        return val;
    }
}
/** Read the tail of a matched element node (tag + properties).
 *  Common prelude (data/position/children) is already consumed by `readMatchedNode`. */
function readElementFromBinary(view, buf, offset, nodeId, resolver, position, childIds, data) {
    let pos = offset;
    // Eager: tagName (almost always accessed by visitors)
    const tagLen = view.getUint16(pos, true);
    pos += 2;
    const tagName = textDecoder.decode(buf.subarray(pos, pos + tagLen));
    pos += tagLen;
    const propsPos = pos;
    // Build node using class (prototype getters, no per-instance defineProperty)
    const node = new WalkElement();
    node.tagName = tagName;
    if (position !== undefined)
        node.position = position;
    if (data !== null)
        node.data = data;
    node._view = view;
    node._buf = buf;
    node._propsPos = propsPos;
    node._childIds = childIds;
    node._resolver = resolver;
    nodeIdMap.set(node, nodeId);
    return node;
}
/** Read a text/comment/raw/expression node from the binary data section. */
const TEXT_NODE_TYPES = {
    2: "text",
    3: "comment",
    5: "raw",
    [HAST_MDX_FLOW_EXPRESSION]: "mdxFlowExpression",
    [HAST_MDX_TEXT_EXPRESSION]: "mdxTextExpression",
    [HAST_MDX_ESM]: "mdxjsEsm",
};
function readTextFromBinary(view, buf, offset, nodeId, nodeType, position, data) {
    const valLen = view.getUint32(offset, true);
    const value = textDecoder.decode(buf.subarray(offset + 4, offset + 4 + valLen));
    const base = { type: TEXT_NODE_TYPES[nodeType], value };
    if (position !== undefined)
        base.position = position;
    if (data !== null)
        base.data = data;
    const node = base;
    nodeIdMap.set(node, nodeId);
    if (nodeType === HAST_MDX_FLOW_EXPRESSION || nodeType === HAST_MDX_TEXT_EXPRESSION) {
        attachParseExpression(node, napiParseExpression);
    }
    else if (nodeType === HAST_MDX_ESM) {
        attachParseExpression(node, napiParseEsm);
    }
    return node;
}
/** Read an MDX JSX element from the binary data section. */
function readMdxJsxFromBinary(view, buf, offset, nodeId, nodeType, resolver, position, childIds, data) {
    let pos = offset;
    // Name
    const nameLen = view.getUint16(pos, true);
    pos += 2;
    const name = nameLen > 0 ? textDecoder.decode(buf.subarray(pos, pos + nameLen)) : null;
    pos += nameLen;
    // Attributes: [kind: u8][nameLen: u16][name][valLen: u32][val]
    const attrCount = view.getUint16(pos, true);
    pos += 2;
    const attributes = [];
    for (let i = 0; i < attrCount; i++) {
        const kind = buf[pos];
        pos += 1;
        const attrNameLen = view.getUint16(pos, true);
        pos += 2;
        const attrName = textDecoder.decode(buf.subarray(pos, pos + attrNameLen));
        pos += attrNameLen;
        const attrValLen = view.getUint32(pos, true);
        pos += 4;
        const attrVal = textDecoder.decode(buf.subarray(pos, pos + attrValLen));
        pos += attrValLen;
        switch (kind) {
            case 0: // BooleanProp
                attributes.push({ type: "mdxJsxAttribute", name: attrName, value: null });
                break;
            case 1: // LiteralProp
                attributes.push({ type: "mdxJsxAttribute", name: attrName, value: attrVal });
                break;
            case 2: // ExpressionProp
                attributes.push({
                    type: "mdxJsxAttribute",
                    name: attrName,
                    value: { type: "mdxJsxAttributeValueExpression", value: attrVal },
                });
                break;
            case 3: // Spread
                attributes.push({ type: "mdxJsxExpressionAttribute", value: attrVal });
                break;
        }
    }
    const typeName = nodeType === HAST_MDX_JSX_ELEMENT ? "mdxJsxFlowElement" : "mdxJsxTextElement";
    const base = { type: typeName, name, attributes };
    if (position !== undefined)
        base.position = position;
    if (data !== null)
        base.data = data;
    nodeIdMap.set(base, nodeId);
    makeLazyChildren(base, childIds, resolver);
    return base;
}
function readMatchedNode(view, buf, offset, nodeId, nodeType, resolver) {
    let pos = offset;
    // Shared prelude (matches serialize_hast_node_inline / serialize_mdast_node_inline):
    //   [data_len: u32][data_bytes][position: 24B][child_count: u16][child_ids: N×u32]
    // Data (JSON), eagerly parsed
    const dataLen = view.getUint32(pos, true);
    pos += 4;
    let data = null;
    if (dataLen > 0) {
        const jsonStr = textDecoder.decode(buf.subarray(pos, pos + dataLen));
        try {
            data = JSON.parse(jsonStr);
        }
        catch {
            /* ignore malformed JSON */
        }
        pos += dataLen;
    }
    const position = readPositionPrefix(view, pos);
    pos += 24;
    const childCount = view.getUint16(pos, true);
    pos += 2;
    const childIds = [];
    for (let i = 0; i < childCount; i++) {
        childIds.push(view.getUint32(pos, true));
        pos += 4;
    }
    // Dispatch to type-specific tail (pos now sits at the type-specific section)
    if (nodeType === HAST_ELEMENT) {
        return readElementFromBinary(view, buf, pos, nodeId, resolver, position, childIds, data);
    }
    else if (nodeType === HAST_TEXT ||
        nodeType === HAST_COMMENT ||
        nodeType === HAST_RAW ||
        nodeType === HAST_MDX_FLOW_EXPRESSION ||
        nodeType === HAST_MDX_TEXT_EXPRESSION ||
        nodeType === HAST_MDX_ESM) {
        return readTextFromBinary(view, buf, pos, nodeId, nodeType, position, data);
    }
    else if (nodeType === HAST_MDX_JSX_ELEMENT || nodeType === HAST_MDX_JSX_TEXT_ELEMENT) {
        return readMdxJsxFromBinary(view, buf, pos, nodeId, nodeType, resolver, position, childIds, data);
    }
    // Fallback: minimal node carrying whatever prelude data we found
    const base = { type: `unknown(${nodeType})` };
    if (position !== undefined)
        base.position = position;
    if (data !== null)
        base.data = data;
    const node = base;
    nodeIdMap.set(node, nodeId);
    return node;
}
// Shared helpers
/**
 * Lazy child materializer, serializes the handle's buffer once when first
 * child is accessed, then materializes children from it via HastReader.
 */
class LazyChildResolver {
    #handle;
    #reader = null;
    constructor(handle) {
        this.#handle = handle;
    }
    #ensure() {
        if (!this.#reader) {
            this.#reader = new HastReader(serializeHandle(this.#handle));
        }
        return this.#reader;
    }
    materializeChildren(childIds) {
        const reader = this.#ensure();
        return childIds.map((id) => {
            const node = materializeHastNode(reader, id);
            this.attachLazyData(node, id);
            return node;
        });
    }
    /** Attach a lazy `data` getter backed by the Rust arena's node_data. */
    attachLazyData(node, nodeId) {
        const handle = this.#handle;
        Object.defineProperty(node, "data", {
            get() {
                const json = napiGetNodeData(handle, nodeId);
                const val = json ? JSON.parse(json) : null;
                Object.defineProperty(this, "data", {
                    value: val,
                    writable: true,
                    enumerable: true,
                    configurable: true,
                });
                return val;
            },
            configurable: true,
            enumerable: true,
        });
    }
}
/** Create a lazy `children` property backed by the handle. */
function makeLazyChildren(node, childIds, resolver) {
    Object.defineProperty(node, "children", {
        get() {
            const children = resolver.materializeChildren(childIds);
            Object.defineProperty(this, "children", {
                value: children,
                writable: true,
                enumerable: true,
                configurable: true,
            });
            return children;
        },
        configurable: true,
        enumerable: true,
    });
}
/** Handle a visitor result (sync).
 *  If the result is the same object as the input node, treat it as a no-op
 *  so that context mutations (e.g. setProperty) are not clobbered. */
function handleVisitResult(result, nodeId, returnBuffer, deferred, originalNode) {
    if (result == null)
        return deferred;
    if (result === originalNode)
        return deferred;
    if (result instanceof Promise) {
        const list = deferred ?? [];
        list.push({ nodeId, promise: result, originalNode });
        return list;
    }
    returnBuffer.replaceRawJson(nodeId, JSON.stringify(markHast(result)));
    return deferred;
}
/**
 * Dispatch matched nodes from a binary match buffer to visitor functions.
 * Returns null if all sync, or an array of deferred promises if any visitor was async.
 */
function dispatchMatches(matchBuf, subs, ctx, returnBuffer, resolver) {
    const matchView = new DataView(matchBuf.buffer, matchBuf.byteOffset, matchBuf.byteLength);
    const matchCount = matchView.getUint32(0, true);
    let deferred = null;
    for (let i = 0; i < matchCount; i++) {
        const indexBase = 4 + i * 12;
        const nodeId = matchView.getUint32(indexBase, true);
        const subIndex = matchBuf[indexBase + 4];
        const dataOffset = matchView.getUint32(indexBase + 6, true);
        const sub = subs[subIndex];
        const node = readMatchedNode(matchView, matchBuf, dataOffset, nodeId, sub.nodeType, resolver);
        const result = sub.visitFn(node, ctx);
        deferred = handleVisitResult(result, nodeId, returnBuffer, deferred, node);
    }
    return deferred;
}
/** Merge return-value + context command buffers and release internals. */
function mergeAndReset(returnBuffer, ctx) {
    const ctxCmdBuf = ctx.getCommandBuffer();
    const ctxBuf = ctxCmdBuf.getBuffer();
    const retBuf = returnBuffer.getBuffer();
    const totalLen = retBuf.length + ctxBuf.length;
    let merged;
    if (totalLen === 0) {
        merged = new Uint8Array(0);
    }
    else {
        merged = new Uint8Array(totalLen);
        merged.set(retBuf, 0);
        merged.set(ctxBuf, retBuf.length);
    }
    returnBuffer.reset();
    ctxCmdBuf.reset();
    return { merged, hasMutations: totalLen > 0 };
}
// Handle-based visitor
/**
 * Walk a handle's arena in Rust, dispatch matched nodes to JS visitor functions,
 * and apply mutations back to the handle. No arena buffers cross NAPI.
 *
 * Returns the number of patches dropped because their target was removed or
 * replaced earlier in the same pass (the caller warns when non-zero), or a
 * Promise of that count if any visitor is async.
 */
export function visitHastHandle(handle, plugin, subs, source, fileURL) {
    const getSource = typeof source === "function" ? source : () => source;
    const ctx = new HastVisitorContextImpl(handle, getSource, fileURL);
    const returnBuffer = new CommandBuffer();
    const resolver = new LazyChildResolver(handle);
    const rustSubs = subs.map((s) => ({ nodeType: s.nodeType, tagFilter: s.tagFilter }));
    const deferred = dispatchMatches(walkHandle(handle, rustSubs), subs, ctx, returnBuffer, resolver);
    if (deferred) {
        return Promise.all(deferred.map((d) => d.promise.then((result) => ({ nodeId: d.nodeId, result, originalNode: d.originalNode })))).then((results) => {
            for (const { nodeId, result, originalNode } of results) {
                if (result != null && result !== originalNode) {
                    returnBuffer.replaceRawJson(nodeId, JSON.stringify(markHast(result)));
                }
            }
            return applyMutations(handle, returnBuffer, ctx);
        });
    }
    return applyMutations(handle, returnBuffer, ctx);
}
/** Returns the number of patches dropped as stranded (0 when none). */
function applyMutations(handle, returnBuffer, ctx) {
    const { merged, hasMutations } = mergeAndReset(returnBuffer, ctx);
    if (hasMutations) {
        return applyCommandsToHandle(handle, merged);
    }
    return 0;
}
